runtime/weak.c

```diff
@@ -278,58 +278,70 @@ CAMLprim value caml_weak_get (value ar, value n)
 static value ephe_get_field_copy (value e, mlsize_t offset)
 {
   CAMLparam1 (e);
-  CAMLlocal2 (res, elt);
-  mlsize_t i, infix_offs = 0;
-  value v; /* Caution: this is NOT a local root. */
-  value f;
-
-  clean_field(e, offset);
-  v = Field(e, offset);
-  if (v == caml_ephe_none) {
-    res = Val_none;
-    goto out;
-  }
+  CAMLlocal4 (res, val, copy, field);
+  mlsize_t i = 0, infix_offs = 0;
+  tag_t copy_tag = 0;
+  mlsize_t copy_size = 0;
+  mlsize_t copy_offs = 0;
+
+  /* Loop in case allocating the copy triggers a GC which modifies the
+   * ephemeron or the value. */
+  while(true) {
+    clean_field(e, offset);
+    val = Field(e, offset);
 
-  /** Don't copy custom_block #7279 */
-  if (Is_block(v) && Tag_val(v) != Custom_tag) {
-    if (Tag_val(v) == Infix_tag) {
-      infix_offs = Infix_offset_val(v);
-      v -= infix_offs;
+    if (val == caml_ephe_none) {
+      res = Val_none;
+      goto out;
     }
-    elt = caml_alloc (Wosize_val(v), Tag_val(v));
 
-    clean_field(e, offset);
-    v = Field(e, offset);
-    if (v == caml_ephe_none) CAMLreturn (Val_none);
-
-    if (Tag_val(v) == Infix_tag) {
-      infix_offs = Infix_offset_val(v);
-      v -= infix_offs;
+    /* Don't copy immediates or custom blocks #7279 */
+    if (!Is_block(val) || Tag_val(val) == Custom_tag) {
+      copy = val;
+      goto some;
     }
 
-    if (Tag_val(v) < No_scan_tag) {
-      caml_domain_state* domain_state = Caml_state;
-      i = 0;
-      if (Tag_val (v) == Closure_tag) {
-        /* Direct copy of the code pointers and closure info fields */
-        i = Start_env_closinfo(Closinfo_val(v));
-        memcpy (Bp_val (elt), Bp_val (v), Bsize_wsize (i));
-      }
-      /* Field-by-field copy and darkening of the remaining fields */
-      for (/*nothing*/; i < Wosize_val(v); i++) {
-        f = Field(v, i);
-        caml_darken (domain_state, f, 0);
-        Store_field(elt, i, f);
-      }
-    } else {
-      memmove (Bp_val(elt), Bp_val(v), Bosize_val(v));
+    if (Tag_val(val) == Infix_tag) {
+      infix_offs = Infix_offset_val(val);
+      val -= infix_offs;
     }
-  } else {
-    Field(e, offset) = elt = v;
+
+    if ((Tag_val(val) == copy_tag) &&
+        (Wosize_val(val) == copy_size) &&
+        (infix_offs == copy_offs))
+      break;
+
+    /* This allocation could provoke a GC, which could change the
+     * header or size of val (e.g. in a finalizer). If it does, go
+     * around the loop and try again. */
+    copy = caml_alloc (Wosize_val(val), Tag_val(val));
+    copy_size = Wosize_val(copy);
+    copy_tag = Tag_val(copy);
+    copy_offs = infix_offs;
+    val = Val_unit;
```

Comment:
On `copy_offs`: if `val` has changed since the last iteration, from an infix closure to a non-infix one (say), but both closures are the same size, then surely we need to keep `copy_offs` to catch that? Oh, but we don't care, because if we pass then we're going to get a copy of `val` and offset it with the infix offset from `val`. I see. OK.

Reply:
I'll add a comment so that future developers won't face the same confusion.